### PR TITLE
Incremental builds.  Fix #194.

### LINKED ISSUE #194: `sphinx-js` invalidates every document / does no dependency tracking, making incremental updates very slow
The version 1.2 release notes say

> Always do full rebuilds; don't leave pages stale when JS code has changed but the RSTs have not.

The relevant commit is https://github.com/mozilla/sphinx-js/commit/89baaaa9e9713048c17349112d48162ba2473cd7.

This choice is expedient but does not scale: it takes several minutes to read the entirety of the documentation for https://github.com/mozilla/gecko-dev, some 4000 sources.  Poking around the Sphinx/autodoc source code, it looks like there is some special support for exactly this type of dependency tracking; the [relevant keyword appears to be `record_dependencies`](https://github.com/sphinx-doc/sphinx/blob/068f802df90ea790f89319094e407c4d5f6c26ff/sphinx/ext/autodoc/directive.py#L154-L157).  It seems plausible that this mechanism could be used in `sphinx-js` to improve the situation.

I will note that with `jsdoc` (and `typedoc`) apparently responsible for actually finding relevant input files, there are some layers of indirection to work around.

### PR DESCRIPTION
**Pull request by ncalexan**

### CODE REVIEW

**Comment by ncalexan** [comment]
@erikrose oh hai!  Are you willing to review?  I can probably arrange for somebody within Mozilla to review if you're not available.

File: sphinx_js/typedoc.py [hunk 1]
```diff
@@ -68,6 +68,25 @@ def _containing_module(self, node) -> Optional[Pathname]:
                 # Found one!
                 return Pathname(make_path_segments(node, self._base_dir))
 
+    def _containing_deppath(self, node) -> Optional[Pathname]:
+        """Return the path pointing to the module containing the given node.
+        The path is absolute or relative to `root_for_relative_js_paths`.
+        Raises ValueError if one isn't found.
+
+        """
+        while True:
```
Comment (lonnen) [inline sphinx_js/typedoc.py:77]: This has some code duplication with `_containing_module()`. Maybe that can be extracted to an iterator that `_containing_module()` and `_containing_deppath()` can use

**Review by lonnen — CHANGES_REQUESTED** [review]

**Comment by lonnen** [comment]
@ncalexan Hello and thanks for the PR. This looks like a very useful feature! I took a pass and had two suggestions

File: sphinx_js/renderers.py [hunk 2]
```diff
@@ -57,39 +57,58 @@ def from_directive(cls, directive, app):
                    content=directive.content,
                    options=directive.options)
 
-    def rst_nodes(self):
-        """Render into RST nodes a thing shaped like a function, having a name
-        and arguments.
-
-        Fill in args, docstrings, and info fields from stored JSDoc output.
+    def get_object(self):
```
[the remark this answers is not in the document]
Comment (lonnen) [reply sphinx_js/renderers.py:60]: `get_object()` is used by this name with this signature in both the jsdoc and typedoc modules in addition to this Analyzer class. They're not descendants of one another but that's enough examples for me to change my mind on this.

retracted

**Comment by lonnen** [comment]
I took this branch and added my own review requests in hopes of getting this landed faster. see also: https://github.com/mozilla/sphinx-js/pull/197

**Review by lonnen — APPROVED** [review]

**Comment by lonnen** [comment]
Merged #197

**Pull request closed without merging**